designs/feature-control.md

```diff
@@ -0,0 +1,85 @@
+Feature Access Control for Unreleased Features
+====================================
+
+What is the problem?
+--------------------
+
+Currently, we cannot push large WIP features to the `master` branch as new builds will include unfinished code. Additionally, development in separate, isolated feature branches cause massive merges resulting in days of work to remedy. A feature access control system will let us develop code and push it to the `master` branch as it's written, while hiding the functionality from the general populace.
+
+What will be added?
+---------------------
+
+We will add a feature access control class to the codebase which will allow developers to gate off code paths and allow easily-configurable access to users.
+
+### Success criteria for the addition
+
+* A feature access control system that grants access to unreleased code for development purposes.
+  * Access should not be dependent on a rebuild--this should be user-configurable.
+  * Flags should be undocumented outside the codebase. Even though features will be included in new builds, they will eminently not be ready for general use.
+  * Individual features should be able to be toggled on/off on-the-fly or only on restart ("session-permanent") in a predictable manner.
```

Comment:
What's the value of the `session-permanent` concept? Why wouldn't toggles be re-evaluated each time?

Reply:
The value would be if the feature would otherwise screw up the functionality of something that's already been loaded, or something that's only loaded once. Essentially, a "session-permanent" feature would be something that when toggled, would ordinarily pop a toast that reads "This setting will take effect on restart"...since we aren't surfacing these fields to the user, we don't have any actions that we can latch onto that can pop a toast like that.

Non session-permanent features would be mainly for isolated features.

Reply:
I wonder if we can start with a simpler concept by making all features "session permanent". The design of the system would then require a VS Code restart if any feature was added to/removed from the feature control configuration.

As an example, when you enable or disable the Python JEDI setting, it toasts with a message to restart VS Code for the setting change to take effect.

Reply:
This will put a hard requirement that every feature is also added to the array of session-permanent features since otherwise we won't know what the initial state is.

I don't think we necessarily make the system more complex by requiring everything to be session-permanent versus having a mix; I think the only win is that features would be more clearly documented (this matters less if we have an enum or per-feature functions, as discussed below).

Reply:
The point would be to do away with the second array serving as "session permanent" in the implementation. All feature flags are considered "disabled" as the initial state unless they are added to the configuration. 

If you're developing against an experimental feature, or trying out a provided fix, you should have to explicitly set yourself up to enable the feature.

Reply:
yup